### PR TITLE
Add warning on hydration mismatch that makes it easy to identify it

### PR DESCRIPTION
**Pull request by salazarm**
## Summary

This warning makes it a lot easier to identify where exactly a hydration mismatch occurs. It looks like this:

```
An error occurred during hydration. expected article got DIV. At: [object HTMLDivElement] Inside: [object HTMLDivElement]
```

I opted to log the HTML element object itself rather than stringifying it to make it much easier for developers to identify where exactly the mismatch is. Otherwise it could still be difficult to identify where it is. Open to any other approaches/suggestions.

## How did you test this change?

jest

### CODE REVIEW

File: packages/react-dom/src/client/ReactDOMHostConfig.js [hunk 1]
```diff
@@ -1077,6 +1081,28 @@ export function errorHydratingContainer(parentContainer: Container): void {
...
 }
 
+export function warnOnHydrationMismatch(
+  fiber: Fiber,
+  parentFiber: Fiber | null,
+  hydratableInstance: HydratableInstance | null,
+) {
+  if (__DEV__) {
+    // eslint-disable-next-line react-internal/warning-args
+    console.warn(
```
Comment (gaearon) [inline packages/react-dom/src/client/ReactDOMHostConfig.js:1091]: should be console.error?

Comment (salazarm) [reply packages/react-dom/src/client/ReactDOMHostConfig.js:1091]: Hmm, not sure, is that the convention?

File: packages/react-dom/src/client/ReactDOMHostConfig.js [hunk 2]
```diff
@@ -1077,6 +1081,28 @@ export function errorHydratingContainer(parentContainer: Container): void {
   }
 }
 
+export function warnOnHydrationMismatch(
+  fiber: Fiber,
```
Comment (sebmarkbage) [inline packages/react-dom/src/client/ReactDOMHostConfig.js:1085]: Fibers should never leak into the DOM host config. It currently leaks in a few places in the event system but that's isolated to that and the plan is to get rid of those synthetic parts eventually anyway.

Think of it as I should be able to write a completely different implementation of Fiber that doesn't even use Fibers and still reuse the host config. Which we likely will.

Comment (sebmarkbage) [reply packages/react-dom/src/client/ReactDOMHostConfig.js:1085]: I noticed that you didn't actually add the first Fiber. The test selectors seems to have done it too. We'll have to go back to reaffirm the abstraction boundary with those too.

**Pull request closed without merging**